### ISSUE TITLE
ONNX export is incorrect on Apple Silicon M1 MPS

### ISSUE THREAD

**Issue by NicolasNierenbergPersyst** [pytorch/pytorch#83230]
### 🐛 Describe the bug

@onnxbot 

On the onnx site they asked me to post this here.

In the specific case of a PyTorch model with a Conv2d and a BatchNorm2d if onnx export is run on an M1 GPU mps model then a subsequent forward using onnx will not match a forward in PyTorch. If the same model is on the CPU or on CUDA then the export will produce a model that matches PyTorch. Note that PyTorch forwards on MPS and on CPU match in this case.

The example below shows my best effort at a minimum case. There needs to be more than one feature on the BatchNorm2d. You have to perform one forward before using eval() on the model so that the BatchNorm does a computation.

```python
import copy
import torch
import torch
from torch import nn
import numpy as np
import os, sys
import onnx
import onnxruntime

y = torch.randn(1, 2, 3, 6)
device = 'cuda'
ygpu = y.to(device=device)
#conv = nn.Conv2d(in_channels=57, out_channels=1, kernel_size=(1,3), stride=2)
gpuconv = nn.Sequential(
nn.Conv2d(2, 2, stride=1, kernel_size=3),
nn.BatchNorm2d(num_features=2)
).to(device=device)
xgpu = gpuconv(ygpu)
gpuconv.eval()
xgpu = gpuconv(ygpu)
conv = copy.deepcopy(gpuconv).to(device='cpu')
x = conv(y)

print(torch.allclose(x, xgpu.to(device='cpu'), rtol=1e-03, atol=1e-05))
torch.onnx.export(conv, y, 'test.onnx', export_params=True)
torch.onnx.export(gpuconv, ygpu, 'testgpu.onnx', export_params=True)
onnx_model = onnx.load('testgpu.onnx')
onnx.checker.check_model(onnx_model)

ort_session = onnxruntime.InferenceSession('testgpu.onnx')

def to_numpy(tensor):
    return np.array(tensor.to('cpu').detach().numpy() if tensor.requires_grad else tensor.cpu().numpy())
```

### Versions

Collecting environment information...
PyTorch version: 1.13.0.dev20220801
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.13 (default, Mar 28 2022, 06:13:39)  [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.13.0.dev20220801
[conda] numpy                     1.22.3           py38h25ab29e_0  
[conda] numpy-base                1.22.3           py38h974a1f5_0  
[conda] pytorch                   1.13.0.dev20220801         py3.8_0    pytorch-nightly

cc @kulinseth @albanD

**Comment by albanD** [comment]
cc @BowenBao is that something that you would expect to work on your end? Or should we disable this with a nice error message for now?

**Comment by BowenBao** [comment]
@albanD We will take a quick look from our end. cc @justinchuby.

@NicolasNierenbergPersyst Is BatchNorm required in order for this mismatch to happen? I wonder if the export conv-bn fusion is not working correctly on M1. Will you observe a mismatch if there is solely Conv or BatchNorm in the model, but not both at the same time?

**Comment by justinchuby** [comment]
I'd be curious to know when the mismatch happens. Do you execute both pytorch and onnxruntime on the m1 gpu? Does onnxruntime produce different results in different environments on the _same_ model, or does pytorch produce a different onnx model in different environments?

**Comment by NicolasNierenbergPersyst** [comment]
It took me a while.  It requires both Cons and BatchNorm and you have to do one forward on the model before doing the export to see the problem.  I assume that causes BatchNorm to do a calculation
Nicolas Nierenberg
CEO
Persyst Development Corporation
420 Stevens Ave., Suite 210
Solana Beach, CA 92075

**Comment by NicolasNierenbergPersyst** [comment]
I don’t think onnxruntime uses the GPU at least I don’t know how to set that or see that.  When I do the export with the error the PyTorch model is on the GPU, if it is on the CPU then the correct model is exported to onnx.  A forward in PyTorch on either the GPU or the CPU produces the same correct result.  For comparison I tried this on COLAB with CUDA and I was able to export to onnx with a model either on the CPU or the GPU (CUDA) with the correct result.

**Comment by justinchuby** [comment]
Could you share (1) what the actual outputs are and how they differ (2) Images of the exported onnx models that produced different numbers by using netron? https://github.com/lutzroeder/netron

**Comment by NicolasNierenbergPersyst** [comment]
@justinchuby I'm happy to upload the two models that are produced, but after that maybe someone else can take it from there. test.onnx is the one created when the PyTorch model is on the CPU and testgpu.onnx is the one created when the PyTorch model is on GPU (M1).  They produce two different results, but the test.onnx result matches the PyTorch forward with the same inputs.
[Archive.zip](https://github.com/pytorch/pytorch/files/9312530/Archive.zip)

**Comment by justinchuby** [comment]
I can see while the structure is the same, the weights are different
![image](https://user-images.githubusercontent.com/11205048/184256620-5271db86-4df9-4a86-ae13-e1d10487bf96.png)

**Comment by NicolasNierenbergPersyst** [comment]
@justinchuby sure, that’s what I would expect.  But they are produced from the same PyTorch model as you can see in the code.

**Comment by justinchuby** [comment]
Pending: #83533

**Comment by justinchuby** [comment]
Current reproduction

For some reason batch norm is not in the onnx graph?

```python
x = torch.tensor([[[[ 1.1227, -0.5545, -0.7871,  0.0841,  0.3714,  0.1612],
          [-0.2052, -0.6623,  1.6033,  2.1245,  1.9609, -1.6847],
          [ 0.1736, -0.7240, -0.0737,  0.3207,  1.5287, -1.3479]],

         [[-0.0267,  1.3100, -0.5756,  0.9288,  1.0432, -0.4871],
          [-0.9110, -0.2054, -1.6963,  0.5995,  1.7877, -2.0397],
          [-0.1205, -1.0103, -2.1805,  1.4364, -1.3835,  0.6537]]]])

torch.manual_seed(0)

device = "mps"
x_m1 = x.to(device=device)
conv_m1 = nn.Sequential(
    nn.Conv2d(2, 2, stride=1, kernel_size=3),
    nn.BatchNorm2d(num_features=2)
).to(device=device)

y_m1 = conv_m1(x_m1)
conv_m1.eval()
eval_m1 = conv_m1(x_m1)

conv = copy.deepcopy(conv_m1).to(device='cpu')
y = conv(x)

print(torch.allclose(y, y_m1.to(device='cpu'), rtol=1e-03, atol=1e-05))
torch.onnx.export(conv, x, 'test.onnx', export_params=True)
torch.onnx.export(conv_m1, x_m1, 'testm1.onnx', export_params=True)
onnx_model = onnx.load('testm1.onnx')
onnx.checker.check_model(onnx_model)

ort_session = onnxruntime.InferenceSession('testm1.onnx')
```

```
In [8]: conv_m1.graph
Out[8]:
graph(%self.1 : __torch__.torch.nn.modules.container.___torch_mangle_5.Sequential,
      %input.1 : Float(1, 2, 3, 6, strides=[36, 18, 6, 1], requires_grad=0, device=mps:0)):
  %_1 : __torch__.torch.nn.modules.batchnorm.___torch_mangle_4.BatchNorm2d = prim::GetAttr[name="1"](%self.1)
  %_0 : __torch__.torch.nn.modules.conv.___torch_mangle_3.Conv2d = prim::GetAttr[name="0"](%self.1)
  %53 : Tensor = prim::CallMethod[name="forward"](%_0, %input.1)
  %54 : Tensor = prim::CallMethod[name="forward"](%_1, %53)
  return (%54)

In [9]: conv.graph
Out[9]:
graph(%self.1 : __torch__.torch.nn.modules.container.___torch_mangle_11.Sequential,
      %input.1 : Float(1, 2, 3, 6, strides=[36, 18, 6, 1], requires_grad=0, device=cpu)):
  %_1 : __torch__.torch.nn.modules.batchnorm.___torch_mangle_10.BatchNorm2d = prim::GetAttr[name="1"](%self.1)
  %_0 : __torch__.torch.nn.modules.conv.___torch_mangle_9.Conv2d = prim::GetAttr[name="0"](%self.1)
  %53 : Tensor = prim::CallMethod[name="forward"](%_0, %input.1)
  %54 : Tensor = prim::CallMethod[name="forward"](%_1, %53)
  return (%54)
```

**Comment by justinchuby** [comment]
@NicolasNierenbergPersyst

I don't think I can reproduce what you have observed with the following script:

```python
import copy

import numpy as np
import onnxruntime
import torch
from torch import nn

y = torch.randn(1, 2, 3, 6)
device = torch.device("mps")
y_m1 = y.to(device=device)
m1conv = nn.Sequential(
    nn.Conv2d(2, 2, stride=1, kernel_size=3), nn.BatchNorm2d(num_features=2)
).to(device=device)
x_m1 = m1conv(y_m1)
m1conv.eval()
x_m1 = m1conv(y_m1)
conv = copy.deepcopy(m1conv).to(device="cpu")
x = conv(y)

print(torch.allclose(x, x_m1.to(device="cpu"), rtol=1e-03, atol=1e-05))
torch.onnx.export(conv, y, "test.onnx", export_params=True)
torch.onnx.export(m1conv, y_m1, "testm1.onnx", export_params=True)

ort_session = onnxruntime.InferenceSession("test.onnx")
ort_cpu = ort_session.run(None, {"input.1": y_m1.cpu().numpy()})
print("cpu", ort_cpu)

ort_session = onnxruntime.InferenceSession("testm1.onnx")
ort_m1 = ort_session.run(None, {"input.1": y_m1.cpu().numpy()})
print("m1", ort_m1)

print("pytorch_cpu", x)
print("pytorch_m1", x_m1)

print("assertion...")
torch.testing.assert_allclose(ort_m1[0], x_m1.to(device="cpu"), rtol=1e-03, atol=1e-05)
torch.testing.assert_allclose(ort_cpu[0], x_m1.to(device="cpu"), rtol=1e-03, atol=1e-05)
print("passed")
```

```
True
cpu [array([[[[-0.7683179 ,  0.37499702,  0.54862154,  1.0949671 ]],

        [[-0.07792199, -0.23450597,  0.32800803, -0.5749837 ]]]],
      dtype=float32)]
m1 [array([[[[-0.7683179 ,  0.37499702,  0.54862154,  1.0949671 ]],

        [[-0.07792207, -0.23450601,  0.328008  , -0.5749838 ]]]],
      dtype=float32)]
pytorch_cpu tensor([[[[-0.7683,  0.3750,  0.5486,  1.0950]],

         [[-0.0779, -0.2345,  0.3280, -0.5750]]]],
       grad_fn=<NativeBatchNormBackward0>)
/opt/homebrew/lib/python3.9/site-packages/torch/_tensor_str.py:103: UserWarning: The operator 'aten::bitwise_and.Tensor_out' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at  /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
  nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
pytorch_m1 tensor([[[[-0.7683,  0.3750,  0.5486,  1.0950]],

         [[-0.0779, -0.2345,  0.3280, -0.5750]]]], device='mps:0',
       grad_fn=<NativeBatchNormBackward0>)
assertion...
passed
```

Do you see anything I am missing?

**Comment by NicolasNierenbergPersyst** [comment]
@justinchuby Hi there. The code from your example code fails on my system, I put the output in the first section below.  I'm assuming that you are running a later build and if I update to the latest nightly then this problem has been resolved.  But to verify could you run the code from the second section?  I cleaned it up so you should be able to just run it and see that all the results are True.  This will confirm that on your system you no longer see the original error.

```
True
cpu [array([[[[ 0.11757168,  0.2158733 , -0.082766  ,  0.19090553]],

        [[-0.82177913,  0.1580452 ,  0.3241165 , -0.75152457]]]],
      dtype=float32)]
m1 [array([[[[ 0.11757167,  0.21587332, -0.08276604,  0.19090553]],

        [[-0.8327295 ,  0.16417861,  0.33314556, -0.76124996]]]],
      dtype=float32)]
pytorch_cpu tensor([[[[ 0.1176,  0.2159, -0.0828,  0.1909]],

         [[-0.8218,  0.1580,  0.3241, -0.7515]]]],
       grad_fn=<NativeBatchNormBackward0>)
pytorch_m1 tensor([[[[ 0.1176,  0.2159, -0.0828,  0.1909]],

         [[-0.8218,  0.1580,  0.3241, -0.7515]]]], device='mps:0',
       grad_fn=<NativeBatchNormBackward0>)
assertion...
/Users/nicolasnierenberg/opt/anaconda3/envs/torch-gpu/lib/python3.8/site-packages/torch/_tensor_str.py:107: UserWarning: The operator 'aten::bitwise_and.Tensor_out' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1659337886150/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Input In [16], in <cell line: 36>()
     33 print("pytorch_m1", x_m1)
     35 print("assertion...")
---> 36 torch.testing.assert_allclose(ort_m1[0], x_m1.to(device="cpu"), rtol=1e-03, atol=1e-05)
     37 torch.testing.assert_allclose(ort_cpu[0], x_m1.to(device="cpu"), rtol=1e-03, atol=1e-05)
     38 print("passed")

File ~/opt/anaconda3/envs/torch-gpu/lib/python3.8/site-packages/torch/testing/_deprecated.py:32, in warn_deprecated.<locals>.outer_wrapper.<locals>.inner_wrapper(*args, **kwargs)
     30 @functools.wraps(fn)
     31 def inner_wrapper(*args: Any, **kwargs: Any) -> Any:
---> 32     return_value = fn(*args, **kwargs)
     33     tail = instructions(name, args, kwargs, return_value) if callable(instructions) else instructions
     34     msg = (head + tail).strip()

File ~/opt/anaconda3/envs/torch-gpu/lib/python3.8/site-packages/torch/testing/_deprecated.py:80, in assert_allclose(actual, expected, rtol, atol, equal_nan, msg)
     77 if rtol is None and atol is None:
     78     rtol, atol = _get_default_rtol_and_atol(actual, expected)
---> 80 torch.testing.assert_close(
     81     actual,
     82     expected,
     83     rtol=rtol,
     84     atol=atol,
     85     equal_nan=equal_nan,
     86     check_device=True,
     87     check_dtype=False,
     88     check_stride=False,
     89     msg=msg or None,
     90 )

    [... skipping hidden 1 frame]

File ~/opt/anaconda3/envs/torch-gpu/lib/python3.8/site-packages/torch/testing/_comparison.py:1093, in assert_equal(actual, expected, pair_types, sequence_types, mapping_types, msg, **options)
   1090     return
   1092 # TODO: compose all metas into one AssertionError
-> 1093 raise error_metas[0].to_error(msg)

AssertionError: Tensor-likes are not close!

Mismatched elements: 4 / 8 (50.0%)
Greatest absolute difference: 0.010950267314910889 at index (0, 1, 0, 0) (up to 1e-05 allowed)
Greatest relative difference: 0.038808420976013246 at index (0, 1, 0, 1) (up to 0.001 allowed)
```

```python
import copy
import torch
from torch import nn
import onnx
import onnxruntime
import numpy as np
y = torch.randn(1, 2, 3, 6)
device = 'mps'
ygpu = y.to(device=device)

gpuconv = nn.Sequential(
            nn.Conv2d(2, 2, stride=1, kernel_size=3),
            nn.BatchNorm2d(num_features=2)
            ).to(device=device)
xgpu = gpuconv(ygpu)
gpuconv.eval()
xgpu = gpuconv(ygpu)
conv = copy.deepcopy(gpuconv).to(device='cpu')
x = conv(y)

print(torch.allclose(x, xgpu.to(device='cpu'),  rtol=1e-03, atol=1e-05))
torch.onnx.export(conv, y, 'test.onnx', export_params=True)
torch.onnx.export(gpuconv, ygpu, 'testgpu.onnx', export_params=True)
onnx_model = onnx.load('testgpu.onnx')
onnx.checker.check_model(onnx_model)



def to_numpy(tensor):
    return np.array(tensor.to('cpu').detach().numpy() if tensor.requires_grad else tensor.cpu().numpy())

# compute ONNX Runtime output prediction using GPU version
ort_session = onnxruntime.InferenceSession('testgpu.onnx')
ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(y)}
ort_outs = np.array(ort_session.run(None, ort_inputs)).squeeze()
torch_out_numpy = to_numpy(x).squeeze()
print(np.allclose(torch_out_numpy,ort_outs, rtol=1e-03, atol=1e-05))

# compute ONNX Runtime output prediction using CPU version
ort_session = onnxruntime.InferenceSession('test.onnx')
ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(y)}
ort_outs = np.array(ort_session.run(None, ort_inputs)).squeeze()
torch_out_numpy = to_numpy(x).squeeze()
print(np.allclose(torch_out_numpy,ort_outs, rtol=1e-03, atol=1e-05))
```

**Comment by NicolasNierenbergPersyst** [comment]
@justinchuby Unfortunately after updating to the latest nightly I still see the error when running your code.  Here is the result again and the environment information.  My system is a MacBook Pro with an Apple M1 Max if that makes any difference.

```

  Model Name:	MacBook Pro
  Model Identifier:	MacBookPro18,2
  Chip:	Apple M1 Max
  Total Number of Cores:	10 (8 performance and 2 efficiency)
  Memory:	64 GB
  System Firmware Version:	7459.141.1
  OS Loader Version:	7459.141.1
  Serial Number (system):	QTHCQWPD2D
  Hardware UUID:	A22F04AD-7D2F-5134-8F3D-2124BCEC1A99
  Provisioning UDID:	00006001-0002184414D0401E
  Activation Lock Status:	Enabled
```

```
True
cpu [array([[[[-0.47732258, -0.5391359 ,  0.48755983,  0.24978456]],

        [[-1.9595768 , -0.3177085 , -1.1043472 ,  0.39007303]]]],
      dtype=float32)]
m1 [array([[[[-0.47732258, -0.5391359 ,  0.48755983,  0.24978456]],

        [[-2.047145  , -0.3372853 , -1.1564994 ,  0.3998062 ]]]],
      dtype=float32)]
pytorch_cpu tensor([[[[-0.4773, -0.5391,  0.4876,  0.2498]],

         [[-1.9596, -0.3177, -1.1043,  0.3901]]]],
       grad_fn=<NativeBatchNormBackward0>)
pytorch_m1 tensor([[[[-0.4773, -0.5391,  0.4876,  0.2498]],

         [[-1.9596, -0.3177, -1.1043,  0.3901]]]], device='mps:0',
       grad_fn=<NativeBatchNormBackward0>)
assertion...
/Users/nicolasnierenberg/opt/anaconda3/envs/torch-gpu/lib/python3.8/site-packages/torch/_tensor_str.py:106: UserWarning: The operator 'aten::masked_select' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1660720283288/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  nonzero_finite_vals = torch.masked_select(
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Input In [1], in <cell line: 36>()
     33 print("pytorch_m1", x_m1)
     35 print("assertion...")
---> 36 torch.testing.assert_allclose(ort_m1[0], x_m1.to(device="cpu"), rtol=1e-03, atol=1e-05)
     37 torch.testing.assert_allclose(ort_cpu[0], x_m1.to(device="cpu"), rtol=1e-03, atol=1e-05)
     38 print("passed")

File ~/opt/anaconda3/envs/torch-gpu/lib/python3.8/site-packages/torch/testing/_deprecated.py:32, in warn_deprecated.<locals>.outer_wrapper.<locals>.inner_wrapper(*args, **kwargs)
     30 @functools.wraps(fn)
     31 def inner_wrapper(*args: Any, **kwargs: Any) -> Any:
---> 32     return_value = fn(*args, **kwargs)
     33     tail = instructions(name, args, kwargs, return_value) if callable(instructions) else instructions
     34     msg = (head + tail).strip()

File ~/opt/anaconda3/envs/torch-gpu/lib/python3.8/site-packages/torch/testing/_deprecated.py:80, in assert_allclose(actual, expected, rtol, atol, equal_nan, msg)
     77 if rtol is None and atol is None:
     78     rtol, atol = _get_default_rtol_and_atol(actual, expected)
---> 80 torch.testing.assert_close(
     81     actual,
     82     expected,
     83     rtol=rtol,
     84     atol=atol,
     85     equal_nan=equal_nan,
     86     check_device=True,
     87     check_dtype=False,
     88     check_stride=False,
     89     msg=msg or None,
     90 )

    [... skipping hidden 1 frame]

File ~/opt/anaconda3/envs/torch-gpu/lib/python3.8/site-packages/torch/testing/_comparison.py:1093, in assert_equal(actual, expected, pair_types, sequence_types, mapping_types, msg, **options)
   1090     return
   1092 # TODO: compose all metas into one AssertionError
-> 1093 raise error_metas[0].to_error(msg)

AssertionError: Tensor-likes are not close!

Mismatched elements: 4 / 8 (50.0%)
Greatest absolute difference: 0.08756768703460693 at index (0, 1, 0, 0) (up to 1e-05 allowed)
Greatest relative difference: 0.061618606204381955 at index (0, 1, 0, 1) (up to 0.001 allowed)
```

```
Collecting environment information...
PyTorch version: 1.13.0.dev20220817
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.13 (default, Mar 28 2022, 06:13:39)  [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.13.0.dev20220817
[pip3] torchaudio==0.13.0.dev20220817
[pip3] torchvision==0.14.0.dev20220817
[conda] numpy                     1.22.3           py38h25ab29e_0  
[conda] numpy-base                1.22.3           py38h974a1f5_0  
[conda] pytorch                   1.13.0.dev20220817         py3.8_0    pytorch-nightly
[conda] torchaudio                0.13.0.dev20220817        py38_cpu    pytorch-nightly
[conda] torchvision               0.14.0.dev20220817        py38_cpu    pytorch-nightly
```

**Comment by NicolasNierenbergPersyst** [comment]
@justinchuby Here is one more clue.  It is somewhat dependent on random factors.  If you use torch.ones to initialize then it doesn't fail.  I was also able to get your code to succeed by adding a specific manual_seed to the beginning. (everything from 0 through 8 failed).

```
torch.manual_seed(9) 
```

**Comment by justinchuby** [comment]
@BowenBao After some further testing, I found that the script runs fine with Pytorch 1.12.1, but fails with nightly. I previously tested with 1.12.1.

**Comment by justinchuby** [comment]
git bisect results:

667607b3abf7854127b74b8c8b2cb2237dd67dcb is the first bad commit
commit 667607b3abf7854127b74b8c8b2cb2237dd67dcb
Author: Kulin Seth <...@apple.com>
Date:   Wed Jul 13 05:54:32 2022 +0000

    [MPS] Reduce the number of command_buf created and improve performance (#81338)
    
    The PR improves performance and reduces the CPU overhead by reducing the number of command buffers created. It uses commit and continue feature in MPS.
    Pull Request resolved: https://github.com/pytorch/pytorch/pull/81338
    Approved by: https://github.com/razarmehr

 aten/src/ATen/mps/MPSStream.h                      | 17 ++++-
 aten/src/ATen/mps/MPSStream.mm                     | 73 +++++++++++++++++++-
 aten/src/ATen/native/mps/OperationUtils.mm         | 45 ++++---------
 aten/src/ATen/native/mps/operations/BinaryOps.mm   |  4 +-
 aten/src/ATen/native/mps/operations/Copy.mm        | 78 ++--------------------
 .../ATen/native/mps/operations/Distributions.mm    |  4 +-
 6 files changed, 110 insertions(+), 111 deletions(-)

**Comment by kulinseth** [comment]
Thanks @justinchuby for isolating the offending commit and the bisect script. Seems like we are missing a COMMIT after the deepcopy.

**Comment by kulinseth** [comment]
@NicolasNierenbergPersyst , the change is merged to master. Please try with today's nightly and see if the issue is fixed.

**Comment by kulinseth** [comment]
Please re-open the issue if you still see it.

**Comment by NicolasNierenbergPersyst** [comment]
@kulinseth Sorry I was on vacation without access.  Yes I can confirm that this fixes the issue.  Thank you!

**Comment by NicolasNierenbergPersyst** [comment]
I don’t think onnxruntime uses the GPU at least I don’t know how to set that or see that.  When I do the export with the error the PyTorch model is on the GPU, if it is on the CPU then the correct model is exported to onnx.  A forward in PyTorch on either the GPU or the CPU produces the same correct result.  For comparison I tried this on COLAB with CUDA and I was able to export to onnx with a model either on the CPU or the GPU (CUDA) with the correct result.


Nicolas Nierenberg
CEO
Persyst Development Corporation
420 Stevens Ave., Suite 210
Solana Beach, CA 92075
